klippy/extras/sx1509.py

```diff
@@ -0,0 +1,283 @@
+class SX1509_endstop:
+    def __init__(self, sx1509, mcu_int_pin, pin_params):
+        self._sxpin = int(pin_params['pin'].split('_')[1])
+        self._sx1509 = sx1509
+        self._mcu_int_pin = mcu_int_pin
+        self._mcu = self._sx1509._mcu
+        self._invert = pin_params['invert']
+        self._pullup = pin_params['pullup']
+        self._mcu.register_config_callback(self._build_config)
+        self._bitmask = 1 << self._sxpin
+        # Wrappers
+        self.get_mcu = self._mcu_int_pin.get_mcu
+        self.add_stepper = self._mcu_int_pin.add_stepper
+        self.get_steppers = self._mcu_int_pin.get_steppers
+        self.home_wait = self._mcu_int_pin.home_wait
+        self.query_endstop = self._mcu_int_pin.query_endstop
+        self.query_endstop_wait = self._mcu_int_pin.query_endstop_wait
+        self.TimeoutError = self._mcu_int_pin.TimeoutError
+    def _build_config(self):
+        self._mcu.add_config_cmd("i2c_set_bits oid=%d reg=%02x data=%04x" % (
+            self._sx1509._oid, REG_DIR & 0xFF, self._bitmask & 0xFFFF))
+        if self._pullup:
+            self._mcu.add_config_cmd("i2c_set_bits oid=%d reg=%02x data=%04x" % (
+                self._sx1509._oid, REG_PULLUP & 0xFF, self._bitmask & 0xFFFF))
+            self._mcu.add_config_cmd("i2c_clear_bits oid=%d reg=%02x data=%04x" % (
+                self._sx1509._oid, REG_PULLDOWN & 0xFF, self._bitmask & 0xFFFF))
+        else:
+            self._mcu.add_config_cmd("i2c_clear_bits oid=%d reg=%02x data=%04x" % (
+                self._sx1509._oid, REG_PULLUP & 0xFF, self._bitmask & 0xFFFF))
+            self._mcu.add_config_cmd("i2c_set_bits oid=%d reg=%02x data=%04x" % (
+                self._sx1509._oid, REG_PULLDOWN & 0xFF, self._bitmask & 0xFFFF))
+    def home_prepare(self):
+        if self._sxpin > 7:
+            senseregister = REG_SENSE_B
+        else:
+            senseregister = REG_SENSE_A
+        shift = (self._sxpin & 0x7) * 2
+        if self._invert:
+            sensitivity = 0x2
+        else:
+            sensitivity = 0x1
+        # Enable interrupts for the pin whilst disabling all the others
+        self._sx1509.write_word(REG_INTERRUPT_MASK, ~self._bitmask & 0xFFFF)
+        # Clear both sense
+        self._sx1509.clear_bits(senseregister & 0xFF, (0x3 << shift), 2)
+        # Actually configure the edge we are looking for
+        self._sx1509.set_bits(senseregister & 0xFF, (sensitivity << shift), 2)
+        # Use the interrupt pin for homing
+        self._mcu_int_pin.home_prepare()
+    def home_start(self, print_time, sample_time, sample_count, rest_time):
+        # Read the current state of the sx_pin to see if it is already triggered
+        # at the start of the home move
+        pinstate = self._sx1509.read_word(REG_DATA) & self._bitmask
+        if (self._invert and not pinstate) or (not self._invert and pinstate):
+            raise homing.EndstopError("Probe triggered prior to movement")
```

Comment:
I'm not sure this would work correctly for the normal home case - it's possible for one to issue a G28 when one axis is already at the endstop.  (Only the second home - after the retract - should verify that the homing operation started with the trigger not set.)

Also, I'm not sure this is needed as the homing.py code verifies that the probe/endstop wasn't triggered prior to movement by verifying that the homing operation resulted in at least some stepper movement.  Shouldn't this mechanism continue to work with the sx519?